File: week15/practice15_skeleton/autocomplete_skeleton.py
# Practice 15. Autocomplete system
import sys
BUILD_TRIE = 'T'
AUTOCOMPLETE = 'A'
ENDS_HERE = '#'

if __name__ == '__main__':
  if len(sys.argv) != 3:
    raise Exception("Correct usage: [program] [input] [output]")
  with open(sys.argv[1], 'r') as inFile:
    lines = inFile.readlines()
  with open(sys.argv[2], 'w') as outFile:
    i = 0
    while i < len(lines):
      words = lines[i].split()
      op = words[0]
      if op == BUILD_TRIE:
        if len(words) != 2:
          raise Exception("BUILD_TRIE: invalid input")
        n = int(words[1])
        strings = []
        while n:
          i += 1
          strings.append(lines[i].strip())
          n -= 1
        # TODO. Build a trie with "strings"
      elif op == AUTOCOMPLETE:
        if len(words) != 2:
          raise Exception("AUTOCOMPLETE: invalid input")
        prefix = words[1]
        # TODO. Run autocomplete with "prefix" and write results into "outFile"
      else:
        raise Exception("Undefined operator")
      i += 1

File: week15/autocomplete/autocomplete_skeleton.cpp
// Practice 15. Autocomplete System
#include <algorithm>
#include <fstream>
#include <iostream>
#include <string>
#include <sstream>
using namespace std;
const char BUILD_TRIE = 'T';
const char AUTOCOMPLETE = 'A';
const int CHAR_SIZE = 128; // Contains ASCII codes for a-z & A-Z.

int main(int argc, char* argv[]) {
  if (argc != 3) {
    cerr<<"Correct usage: [program] [input] [output]"<<endl;
    exit(1);
  }
  ifstream inFile(argv[1]);
  ofstream outFile(argv[2]);
  string line;
  while (getline(inFile, line))
  {
    char op = line[0];
    istringstream iss(line.substr(2));
    vector<string> words;
    string prefix;
    int n;
    switch(op)
    {
      case BUILD_TRIE: 
        if (!(iss >> n)) {
          cerr<<"BUILD_TRIE: invalid input"<<endl;
          exit(1);
        }
        for (int i = 0; i < n; ++i) {
          if (!getline(inFile, line)) {
            cerr<<"BUILD_TRIE: invalid input"<<endl;
            exit(1);
          }
          words.push_back(line);
        }
        // TODO. Bulid a trie with the stings in vector "words"
        break;
      case AUTOCOMPLETE:
        if (!(iss >> prefix)) {
          cerr<<"AUTOCOMPLETE: invalid input"<<endl;
          exit(1);
        }
        // TODO. Run autocomplete with "prefix" and write result into "outFile"
        break;
      default:
        cerr<<"Undefined operator"<<endl;
        exit(1);
    }
  }
  outFile.close();
  inFile.close();
}
